pkg/importer/language.go

```diff
@@ -64,20 +68,100 @@ func (e *DiagnosticsError) String() string {
 
 func removeDuplicatePathedValues(pathedValues []PathedLiteralValue) []PathedLiteralValue {
 	uniqueValues := make([]PathedLiteralValue, 0)
-	occurrences := make(map[string]int)
+	identityOccurrences := make(map[string]int)
+	outputOccurrences := make(map[string]int)
 	for _, pathedValue := range pathedValues {
-		occurrences[pathedValue.Value]++
+		if pathedValue.Identity {
+			identityOccurrences[pathedValue.Value]++
+		} else {
+			outputOccurrences[pathedValue.Value]++
+		}
 	}
 
 	for _, pathedValue := range pathedValues {
-		if occurrences[pathedValue.Value] > 1 {
+		if pathedValue.Identity && identityOccurrences[pathedValue.Value] > 1 {
+			// an identity value that has occurred multiple times is not unique
+			continue
+		}
+
+		if !pathedValue.Identity && outputOccurrences[pathedValue.Value] > 1 {
+			// an output value that has occurred multiple times is not unique
 			continue
 		}
+
 		uniqueValues = append(uniqueValues, pathedValue)
 	}
 	return uniqueValues
 }
 
+func nextPropertyPath(path hcl.Traversal, key hcl.Traverser) hcl.Traversal {
+	return append(path, key)
+}
+
+func createPathedValue(
+	root string,
+	property resource.PropertyValue,
+	currentPath hcl.Traversal,
+) []PathedLiteralValue {
+	pathedLiteralValues := make([]PathedLiteralValue, 0)
+	if property.IsNull() {
+		return pathedLiteralValues
+	}
+
+	if property.IsString() {
```

Comment:
What about non-string values, such as `IsNumber()`?

Reply:
Also, what about more complex values, like arrays and objects that are deeply equal? e.g. a resource has an output that's an array of strings and another resource has the same array of strings as an input?

Reply:
> What about non-string values, such as IsNumber()?

I didn't account for numeric values, nor booleans because I am afraid of _overguessing_ dependencies that have nothing to do with each other. For example `instanceSize: 2` on a resource A and `retryCount: 2` on a resource B would infer a dependency that is not necessarily what users wanted. Felt like starting with strings would make the most sense since AFAIK all IDs are strings and references between resources use a lot of them (i.e. `resourceGroup.name` in azure-native, `vpc.id` in aws, etc.)

> Also, what about more complex values, like arrays and objects that are deeply equal? e.g. a resource has an output that's an array of strings and another resource has the same array of strings as an input?

First objects: for these we do not need to check for deep equality because their elements will be references. For example, consider the data here:
```
resourceA {
  // outputs
  config {
      first = "A"
      second = "B"
  }
}

resourceB {
  // inputs
  config {
      first = "A"
      second = "B"
  }
}
```
We could infer that `resourceB.config = resourceA.config` but chances are these are not the same type in the SDK so this might not work. That said, this is not needed in the first place because the current logic would infer dependency by values:
```
resourceB {
  config {
      first = resourceA.config.first
      second = resourceB.config.second
  }
}
```
This version is better handled by program-gen implementations.

As for arrays, I am not sure. The current logic would index each element where it is referenced:
```
resourceA {
  // outputs
  data = ["first", "second"]
}

resourceB {
  // inputs
  data = ["first", "second"]
}
```
Inferred:
```
resourceB {
  // inputs
  data = [resourceA.data[0], resourceA.data[1]]
}
```
so it will work even though it looks a bit ugly